### PR TITLE
Updates for portability and CI 

### PR DESCRIPTION
**Pull request by johnsonjh**
* Allow building with D ImportC
  * Tested working with DMD64 v2.105.0b1 and GDC 13.2.1.
* Add support for Clang-CL
  * Tested with Clang-CL 16.0.6 on Linux using MSVC 17.6 w/Windows 10 SDK headers. The only changes needed are not enabling SAL, and using lowercase (or the exact on-disk capitalization) for library names specified via pragma.
* Add support for OrangeC (≥6.0.70.93)
  * Add a warning for unsupported OrangeC versions.
  * Detect if tests are running under Wine, if so,
    * Report the Wine version, and,
    * Adjust some paths accordingly to avoid false positive failures.
* Don't include `dlfcn.h` if not building plugins
  * Since `dlfcn.h` might not exist (and might be the reason why libsir is being built without plugins); do not include `dlfcn.h` for `SIR_NO_PLUGINS` builds.

### CODE REVIEW

File: tests/tests.c [hunk 1]
```diff
@@ -1431,10 +1436,43 @@ bool sirtest_os_log(void) {
...
+#endif
+    return NULL;
+}
+
 bool sirtest_filesystem(void) {
     INIT(si, SIRL_ALL, 0, 0, 0);
     bool pass = si_init;
 
+    /* Wine version */
+    printf("\t_sirtest_wine: '%s'\n", PRN_STR(_sirtest_wine()));
```
Comment (aremmell) [inline tests/tests.c:1474]: Maybe have this say Running under Wine: %s...", (NULL != winever) ? "yes" : "no"

File: tests/tests.c [hunk 2]
```diff
@@ -1618,8 +1656,8 @@ bool sirtest_filesystem(void) {
 # endif
         {"/dev", true},
 #else /* __WIN__ */
-        {"\\Windows", true},
-        {"\\Program Files", true},
+        {"C:\\Windows", true},
```
Comment (aremmell) [inline tests/tests.c:1659]: If this isn't needed for Wine or something, \ is translated to "root" or `/` `C:\` (or whatever disk Windows is installed on, which isn't guaranteed to be named C).

**Review by aremmell — CHANGES_REQUESTED** [review]
Sorry, there's a lot to cover here...

File: include/sir/platform.h [hunk 3]
```diff
@@ -26,7 +26,7 @@
 #ifndef _SIR_PLATFORM_H_INCLUDED
 # define _SIR_PLATFORM_H_INCLUDED
 
-# if defined(_MSC_VER)
+# if defined(_MSC_VER) && !defined(__clang__)
```
[the remark this answers is not in the document]
Comment (johnsonjh) [reply include/sir/platform.h:29]: Purely for building with mainline Clang-CL.  

MSVC always defines `_MSC_VER` but Clang's Clang-CL (https://clang.llvm.org/docs/MSVCCompatibility.html) is differentiated by defining both `__clang__` and `_MSC_VER`. 

Had to add that here so it doesn't try to use the SAL stuff here (only real MSVC supports it).

File: include/sir.h [hunk 4]
```diff
@@ -505,8 +505,9 @@ bool sir_settextstyle(sir_level level, sir_textattr attr, sir_textcolor fg,
  * @see ::sir_settextstyle
  * @see ::default
  *
- * @returns `true` if succcessfully reset, `false` otherwise. Use ::sir_geterror
- *          to obtain information about any error that may have occurred.
+ * @returns bool `true` if succcessfully reset, `false` otherwise. Use
```
[the remark this answers is not in the document]
Comment (johnsonjh) [reply include/sir.h:508]: Will fix.

File: src/sirfilesystem.c [hunk 5]
```diff
@@ -156,10 +156,20 @@ char* _sir_getcwd(void) {
     return cur;
 # endif
 #else /* __WIN__ */
+# if defined(__ORANGEC__)
+    char cur[SIR_MAXPATH + 1];
```
[the remark this answers is not in the document]
Comment (johnsonjh) [reply src/sirfilesystem.c:160]: 1. No, it's using it's own thing for that.  Ugh.
2. It should ...
3. I'll remove the +1, thanks.

File: include/sir/platform.h [hunk 6]
```diff
@@ -282,7 +286,9 @@ _set_thread_local_invalid_parameter_handler(
 # define SIR_MAXHOST 256
 
 # if !defined(__WIN__)
-#  include <dlfcn.h>
+#  if !defined(SIR_NO_PLUGINS)
```
[the remark this answers is not in the document]
Comment (johnsonjh) [reply include/sir/platform.h:289]: Just this one now, since there were other changes that would have caused conflicts and need rebasing, so I moved it all to this.

File: tests/tests.c [hunk 7]
```diff
@@ -1431,10 +1436,43 @@ bool sirtest_os_log(void) {
 #endif
 }
 
+char *_sirtest_wine(void) {
+#if defined(__WIN__)
+# if defined(_MSC_VER) && !defined(__clang__)
+#  pragma warning(push)
+#  pragma warning(disable : 4152)
+# endif
+    static const char *(CDECL *_p_wine_get_version)(void);
```
[the remark this answers is not in the document]
Comment (johnsonjh) [reply tests/tests.c:1445]: Done.

File: tests/tests.c [hunk 8]
```diff
@@ -1431,10 +1436,43 @@ bool sirtest_os_log(void) {
...
+    HMODULE _h_ntdll;
+
+    _h_ntdll = GetModuleHandle("ntdll.dll");
+    if (_h_ntdll != NULL) {
+# if defined(__GNUC__)
+#  pragma GCC diagnostic push
+#  pragma GCC diagnostic ignored "-Wpedantic"
+# endif
+        _p_wine_get_version = (void *)GetProcAddress(_h_ntdll, "wine_get_version");
```
[the remark this answers is not in the document]
Comment (johnsonjh) [reply tests/tests.c:1454]: Done.

File: src/sirfilesystem.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (aremmell) [reply src/sirfilesystem.c:160]: Okay, if it can use Win32 API, give me 10 mins and I'll write a version that will work for both

File: tests/tests.c [hunk 9]
```diff
@@ -1431,10 +1436,43 @@ bool sirtest_os_log(void) {
...
+# if defined(__GNUC__)
+#  pragma GCC diagnostic pop
+# endif
+        char *wine_version = (char *)_p_wine_get_version();
+        if (wine_version)
+            return wine_version;
```
[the remark this answers is not in the document]
Comment (johnsonjh) [reply tests/tests.c:1460]: Yes, intentional, as wine will always have a "ntdll.dll" loaded (and you can't unload it.)  https://learn.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulehandlea says "Retrieves a module handle for the specified module. The module must have been loaded by the calling process.".  So I won't explicitly load or unload it, so, on real Windows, the handle is going to be NULL and the block won't even run, but in the case that ntdll.dll is already mapped, there won't be any "wine_get_version" function to run.

File: tests/tests.c (hunk 2 above)
Comment (johnsonjh) [reply tests/tests.c:1659]: Ah, see, this is failing under Wine, it was always going for the path based on the current working directory, and under wine that is by default "Z:" (`Z:` maps to Unix `/`), so it was looking for `Z:\Windows` and no good.  There is always a `C:\Windows` on Wine.

Comment (johnsonjh) [reply tests/tests.c:1659]: @aremmell Not sure how you'd want to handle this one then.

File: tests/tests.c [hunk 10]
```diff
@@ -1431,10 +1436,43 @@ bool sirtest_os_log(void) {
 #endif
 }
 
+char *_sirtest_wine(void) {
```
[the remark this answers is not in the document]
Comment (johnsonjh) [reply tests/tests.c:1439]: Done.

File: tests/tests.c (hunk 1 above)
Comment (aremmell) [reply tests/tests.c:1474]: That didn't make sense. Maybe make it say "Wine version: <version> or <not under Wine>"

File: tests/tests.c (hunk 2 above)
Comment (aremmell) [reply tests/tests.c:1659]: I will write a new getcwd. Gimmie 10 mins. If that one still returns Z: then we can leave it C:\

File: tests/tests.c (hunk 1 above)
Comment (johnsonjh) [reply tests/tests.c:1474]: We'll try `printf("\tRunning under Wine: %s\n", sirtest_get_wineversion() ? sirtest_get_wineversion() : "no");`

```
(1/1) 'filesystem'...                                            
                                
        Running under Wine: no
```

Or else it'll show Running under Wine: 8.11 ...

File: include/sir/platform.h [hunk 11]
```diff
@@ -82,6 +82,10 @@
 #    endif
 #   endif
 #  endif
+#  if defined(__IMPORTC__)
```
[the remark this answers is not in the document]
Comment (johnsonjh) [reply include/sir/platform.h:85]: "ImportC is an implementation of ISO/IEC 9899:2011, which will be referred to as C11; ImportC is a C compiler embedded into the D implementation. It enables direct importation of C files, without needing to manually prepare a D file corresponding to the declarations in the C file. It directly compiles C files into modules that can be linked in with D code to form an executable. It can be used as a C compiler to compile and link 100% C programs."

TL;DR - It's a C11 compiler included with the the Dlang compiler. =)

File: tests/tests.c (hunk 2 above)
Comment (johnsonjh) [reply tests/tests.c:1659]: I'll add OrangeC to the CI/CD job too, so we can see what it does in the test.

File: tests/tests.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (aremmell) [reply tests/tests.c:1460]: Sorry, I misread this as LoadLibrary, not GetModuleHandle

File: tests/tests.c [hunk 12]
```diff
@@ -1431,10 +1436,44 @@ bool sirtest_os_log(void) {
...
+    static const char* (CDECL* _p_wine_get_version)(void);
+    HMODULE _h_ntdll;
+
+    _h_ntdll = GetModuleHandle("ntdll.dll");
+    if (_h_ntdll != NULL) {
+# if defined(__GNUC__)
+#  pragma GCC diagnostic push
+#  pragma GCC diagnostic ignored "-Wpedantic"
+# endif
+        _p_wine_get_version = (void *)GetProcAddress(_h_ntdll, "wine_get_version");
```
Comment (aremmell) [inline tests/tests.c:1454]: Is this the line that triggers warning 4152? Check out the example here. I believe if you do it like this, you won't get a warning (use STDCALL instead of CDECL, and typedef that function signature, then cast to that type not void*)

https://learn.microsoft.com/en-us/cpp/build/getprocaddress?view=msvc-170

File: tests/tests.c [hunk 13]
```diff
@@ -1431,10 +1436,44 @@ bool sirtest_os_log(void) {
...
+# if defined(__GNUC__)
+#  pragma GCC diagnostic pop
+# endif
+        char *wine_version = (char *)_p_wine_get_version();
```
Comment (aremmell) [inline tests/tests.c:1458]: Plz asterisks to left of space 😬

**Review by aremmell — CHANGES_REQUESTED** [review]

File: tests/tests.c (hunk 12 above)
Comment (johnsonjh) [reply tests/tests.c:1454]: I'll check it out shortly, but it needs to work on more than just MSVC ... so pretty sure the warning still happens elsewhere, let me see ...

File: include/sir/platform.h [hunk 14]
```diff
@@ -446,6 +451,24 @@ typedef BOOL(CALLBACK* sir_once_fn)(PINIT_ONCE, PVOID, PVOID*);
...
 # endif
 
+# if defined(__WIN__) && defined(__ORANGEC__)
+#  if defined(__ORANGEC_MAJOR__) && defined(__ORANGEC_MINOR__) && defined(__ORANGEC_PATCHLEVEL__)
+#   if __ORANGEC_MAJOR__ <= 6
+#    if __ORANGEC_MINOR__ <= 70
+#     if __ORANGEC_PATCHLEVEL__ <= 92
+#      if !defined(ORANGEC_VERSION_WARNING)
+#       warning OrangeC versions before 6.0.70.93 are unsupported.
+#       define ORANGEC_VERSION_WARNING
```
Comment (aremmell) [inline include/sir/platform.h:461]: Shouldn't we just #error here? what's the point of warning and continuing?

Comment (johnsonjh) [reply include/sir/platform.h:461]: Earlier versions can be made to work, but they're missing some functionality like strnlen - up to you. :)

File: tests/tests.c (hunk 12 above)
Comment (johnsonjh) [reply tests/tests.c:1454]: @aremmell I'm stuck downstairs for a bit away from a real keyboard so if you want to try it and we'll see what CI thinks, feel free, unless I get to it first

**Review by aremmell — APPROVED** [review]
